File: rust/src/solutions/remove_element.rs
// LeetCode 27. Remove Element in Rust

use super::Solution;

impl Solution {
    pub fn remove_element(nums: &mut Vec<i32>, val: i32) -> i32 {
        for index in (0..nums.len()).rev() {
            if nums[index] == val {
                nums.swap_remove(index);
            }
        }
        nums.len() as i32
    }
}

#[cfg(test)]
mod tests {
    use super::Solution;

    #[test]
    fn example1() {
        let mut nums = vec![3, 2, 2, 3];
        let val = 3;
        let ans = 2;
        assert_eq!(Solution::remove_element(&mut nums, val), ans);
        nums.sort();
        assert_eq!(nums, [2, 2]);
    }

    #[test]
    fn example2() {
        let mut nums = vec![0, 1, 2, 2, 3, 0, 4, 2];
        let val = 2;
        let ans = 5;
        assert_eq!(Solution::remove_element(&mut nums, val), ans);
        nums.sort();
        assert_eq!(nums, [0, 0, 1, 3, 4]);
    }
}

File: cpp/src/solutions/remove_element.cpp
// LeetCode 27. Remove Element

#include "cpp/src/common/common.h"

class Solution {
 public:
  int removeElement(vector<int>& nums, int val) {
    for (auto iter = nums.rbegin(); iter != nums.rend(); iter++) {
      if (*iter == val) {
        std::swap(*iter, *nums.rbegin());
        nums.pop_back();
      }
    }
    return nums.size();
  }
};

TEST(removeElement, example1) {
  vector<int> nums = {3, 2, 2, 3};
  int val = 3;
  int ans = 2;
  EXPECT_EQ(Solution().removeElement(nums, val), ans);
  std::sort(nums.begin(), nums.end());
  EXPECT_EQ(nums, vector<int>({2, 2}));
}

TEST(removeElement, example2) {
  vector<int> nums = {0, 1, 2, 2, 3, 0, 4, 2};
  int val = 2;
  int ans = 5;
  EXPECT_EQ(Solution().removeElement(nums, val), ans);
  std::sort(nums.begin(), nums.end());
  EXPECT_EQ(nums, vector<int>({0, 0, 1, 3, 4}));
}
